crates/constraint-vm/src/access/dec_vars.rs

```diff
@@ -0,0 +1,181 @@
+use super::test_utils::ops;
+use super::*;
+use crate::exec_ops;
+use crate::test_util::test_empty_keys;
+use crate::test_util::test_transient_data;
+use asm::Op;
+use essential_constraint_asm as asm;
+use essential_types::ContentAddress;
+use essential_types::PredicateAddress;
+use test_case::test_case;
+use test_utils::assert_err;
+use test_utils::assert_stack_ok;
+
+#[test_case(
+    &[0, 0, 2], &[&[3, 99], &[4, 61, 100]] => using assert_stack_ok(&[3, 99])
+    ; "Sanity"
+)]
+#[test_case(
+    &[1, 0, 2], &[&[3, 99], &[4, 61, 100]] => using assert_stack_ok(&[4, 61])
+    ; "slot_ix 1"
+)]
+#[test_case(
+    &[1, 1, 1], &[&[3, 99], &[4, 61, 100]] => using assert_stack_ok(&[61])
+    ; "slot_ix 1 index 1 len 1"
+)]
+#[test_case(
+    &[1, 1, 0], &[&[3, 99], &[4, 61, 100]] => using assert_stack_ok(&[])
+    ; "empty"
+)]
+#[test_case(
+    &[], &[&[3, 99], &[4, 61, 100]] =>
+    using assert_err!(OpError::Access(AccessError::MissingArg(MissingAccessArgError::DecVarLen)))
+    ; "missing len"
+)]
+#[test_case(
+    &[1], &[&[3, 99], &[4, 61, 100]] =>
+    using assert_err!(OpError::Access(AccessError::MissingArg(MissingAccessArgError::DecVarValueIx)))
+    ; "missing value_ix"
+)]
+#[test_case(
+    &[0, 1], &[&[3, 99], &[4, 61, 100]] =>
+    using assert_err!(OpError::Access(AccessError::MissingArg(MissingAccessArgError::DecVarSlotIx)))
+    ; "missing slot_ix"
+)]
+#[test_case(
+    &{
+        let mut v = vec![1; Stack::SIZE_LIMIT];
+        *v.get_mut(Stack::SIZE_LIMIT - 1).unwrap() = 5;
+        v
+    },  &[&[], &[3; 6]] =>
+    using assert_err!(OpError::Stack(StackError::Overflow))
+    ; "values over flow the stack"
+)]
+#[test_case(
+    &[-1, 0, 1], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::DecisionSlotIxOutOfBounds(-1)))
+    ; "negative slot_ix"
+)]
+#[test_case(
+    &[1, -1, 1], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::InvalidAccessRange))
+    ; "negative value_ix"
+)]
+#[test_case(
+    &[0, 0, -1], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::InvalidAccessRange))
+    ; "negative len"
+)]
+#[test_case(
+    &[1, 0, 1], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::DecisionSlotIxOutOfBounds(1)))
+    ; "slot_ix out of bounds"
+)]
+#[test_case(
+    &[0, 1, 1], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::DecisionValueRangeOutOfBounds(1, 2)))
+    ; "value ix out of bounds"
+)]
+#[test_case(
+    &[0, 0, 2], &[&[3]] =>
+    using assert_err!(OpError::Access(AccessError::DecisionValueRangeOutOfBounds(0, 2)))
+    ; "len out of bounds"
+)]
```

Comment:
What does `#[test_case]` provide over just calling this from different test fns? E.g. looks like this could just be

```suggestion
#[test]
fn len_out_of_bounds() {
    let err = test_dec_var(&[0, 0, 2], &[&[3]]).unwrap_err();
    assert_eq!(err, OpError::Access(AccessError::DecisionValueRangeOutOfBounds(0, 2)));
}
```
I guess the macro is slightly less to type, but maintainers then need to learn this macro :thinking: I guess it's still pretty readable though.

Reply:
Yeh I think it's much more readable and creates a clear mapping from domain to expected co-domain. The macro is incredibly simple. Just `input => expected output`.
I'm open to removing the macro though if it adds confusion.

Reply:
> The macro is incredibly simple

Yeah I thought it was simple and intuitive, then I read `=> using assert_err`, then I thought the macro was introducing this `assert_err!` macro. Then I realised we are the ones introducing it, and that it just introduces `using`. Then I tried to infer what `using` means. It looks like it automagically interpolates the args of the following macro and then inserts the result of the function as the first argument? How else could `assert_err` be getting the result of the function? :thinking: Then I realised, it looks like our added `assert_*` macros/fns are not like normal `assert!`s - they actually return functions. So the macro or function provided to `using` needs to *return* a function that the macro can call in its generated code. So in `using <expr>`, `<expr>` should result in a function value.

At some point during this I was wondering why they're not just `#[test]` fns lol. But I get it now and I'm happy to keep them if you find them handy.

Reply:
Maybe I can add some docs to make things clearer